sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteArithmeticFiltersOnIntegralColumn.scala

```diff
@@ -0,0 +1,157 @@
+  }
+
+  private def transformLeft(
+      bc: BinaryComparison,
+      left: BinaryArithmetic,
+      right: Expression): Expression = {
+    left match {
+      case Add(ar: AttributeReference, e) if e.foldable && isOptSafe(Subtract(right, e)) =>
+        bc.makeCopy(Array(ar, Subtract(right, e)))
+      case Add(e, ar: AttributeReference) if e.foldable && isOptSafe(Subtract(right, e)) =>
+        bc.makeCopy(Array(ar, Subtract(right, e)))
+      case Subtract(ar: AttributeReference, e) if e.foldable && isOptSafe(Add(right, e)) =>
+        bc.makeCopy(Array(ar, Add(right, e)))
+      case Subtract(e, ar: AttributeReference) if e.foldable && isOptSafe(Subtract(e, right)) =>
+        bc.makeCopy(Array(Subtract(e, right), ar))
+      case _ => bc
+    }
+  }
+
+  private def transformRight(
+      bc: BinaryComparison,
+      left: Expression,
+      right: BinaryArithmetic): Expression = {
+    right match {
+      case Add(ar: AttributeReference, e) if e.foldable && isOptSafe(Subtract(left, e)) =>
+        bc.makeCopy(Array(Subtract(left, e), ar))
+      case Add(e, ar: AttributeReference) if e.foldable && isOptSafe(Subtract(left, e)) =>
+        bc.makeCopy(Array(Subtract(left, e), ar))
+      case Subtract(ar: AttributeReference, e) if e.foldable && isOptSafe(Add(left, e)) =>
+        bc.makeCopy(Array(Add(left, e), ar))
+      case Subtract(e, ar: AttributeReference) if e.foldable && isOptSafe(Subtract(e, left)) =>
+        bc.makeCopy(Array(ar, Subtract(e, left)))
+      case _ => bc
+    }
+  }
+
+  private def isDataTypeSafe(dataType: DataType): Boolean = dataType match {
+    case ByteType | ShortType | IntegerType | LongType => true
+    case _ => false
+  }
+
+  private def isOptSafe(e: BinaryArithmetic): Boolean = {
+    val leftVal = e.left.eval(EmptyRow)
+    val rightVal = e.right.eval(EmptyRow)
+
+    e match {
+      case Add(_, _) =>
+        e.dataType match {
+          case ByteType =>
+            isAddSafe(leftVal, rightVal, Byte.MinValue, Byte.MaxValue)
+          case ShortType =>
+            isAddSafe(leftVal, rightVal, Short.MinValue, Short.MaxValue)
+          case IntegerType =>
+            isAddSafe(leftVal, rightVal, Int.MinValue, Int.MaxValue)
+          case LongType =>
+            isAddSafe(leftVal, rightVal, Long.MinValue, Long.MaxValue)
+          case _ => false
+        }
+
+      case Subtract(_, _) =>
+        e.dataType match {
+          case ByteType =>
+            isSubtractSafe(leftVal, rightVal, Byte.MinValue, Byte.MaxValue)
+          case ShortType =>
+            isSubtractSafe(leftVal, rightVal, Short.MinValue, Short.MaxValue)
+          case IntegerType =>
+            isSubtractSafe(leftVal, rightVal, Int.MinValue, Int.MaxValue)
+          case LongType =>
+            isSubtractSafe(leftVal, rightVal, Long.MinValue, Long.MaxValue)
+          case _ => false
+        }
+
+      case _ => false
+    }
+  }
+
+  private def isAddSafe[T](left: Any, right: Any, minValue: T, maxValue: T)(
```

Comment:
I don't see the need for this and the next. As of now, we are not handling overflows with integers (you can see https://github.com/apache/spark/pull/21599 is still open). So I think we can get rid of these checks. It may be worth, though, to add a comment (like a TODO) in order to remind that this issue can arise

Reply:
With some cases, it doesn't necessarily cause overflow if we don't rewrite it. So there's potential inconsistency again.

Reply:
The check here is to make sure if overflow may occur after rewrite, it will not rewrite this expression.